day_two_guide/topics/proc_backing-up-etcd.adoc

```diff
@@ -60,55 +61,73 @@ Before backing up etcd:
 * Ensure that the {product-title} API service is running.
 * Ensure connectivity with the etcd cluster (port 2379/tcp).
 * Ensure the proper certificates to connect to the etcd cluster.
-* Ensure `go` is installed.
 
 ifeval::["{context}" == "environment-backup"]
-. To ensure the etcd cluster is working, check its health.
-** Run the following command:
+* Ensure that the etcd cluster is working by checking its health:
+** Check the health of the endpoints:
++
+// TODO: is this okay to update to use etcdctl3 (since etcdctl doesn't exist on master node)? Remove ETCDCTL_API=3?
 +
 ----
-# ETCDCTL_API=3 etcdctl --cert="/etc/etcd/peer.crt" \
+# ETCDCTL_API=3 etcdctl3 --cert="/etc/etcd/peer.crt" \
           --key=/etc/etcd/peer.key \
           --cacert="/etc/etcd/ca.crt" \
-          --endpoints="https://*master-0.example.com*:2379,\
-            https://*master-1.example.com*:2379,\
-            https://*master-2.example.com*:2379"
-            endpoint health
+          --endpoints="https://master-0.example.com:2379,https://master-1.example.com:2379,https://master-2.example.com:2379" \// <1>
+          endpoint health
+----
+<1> Replace these values with your cluster's endpoints.
++
+.Example Output
+----
 https://master-0.example.com:2379 is healthy: successfully committed proposal: took = 5.011358ms
 https://master-1.example.com:2379 is healthy: successfully committed proposal: took = 1.305173ms
 https://master-2.example.com:2379 is healthy: successfully committed proposal: took = 1.388772ms
 ----
 
-. Check the member list.
+** Check the member list.
 +
 ----
 # etcdctl3 member list
+----
++
+.Example Output
+----
 2a371dd20f21ca8d, started, master-1.example.com, https://192.168.55.12:2380, https://192.168.55.12:2379
 40bef1f6c79b3163, started, master-0.example.com, https://192.168.55.8:2380, https://192.168.55.8:2379
 95dc17ffcce8ee29, started, master-2.example.com, https://192.168.55.13:2380, https://192.168.55.13:2379
 ----
 endif::[]
 
 ifeval::["{context}" == "day-two-host-level-tasks"]
-. To ensure the etcd cluster is working, check its health.
+* Ensure that the etcd cluster is working by checking its health:
+** Check the health of the endpoints:
++
+// TODO: is this okay to update to use etcdctl3 (since etcdctl doesn't exist on master node)? Remove ETCDCTL_API=3?
 +
 ----
-# ETCDCTL_API=3 etcdctl --cert="/etc/etcd/peer.crt" \
+# ETCDCTL_API=3 etcdctl3 --cert="/etc/etcd/peer.crt" \
```

Comment:
@geliu2016 / @hexfusion same question on if it's okay to use `etcdctl3` or not here. We're just checking the health as a prerequisite to taking the backup.

Reply:
@hexfusion @bergerhoffer , I retest it today, but there is some issue about this cluster:

# ETCDCTL_API=3 etcdctl3 --cert="/etc/etcd/peer.crt"     --key=/etc/etcd/peer.key    --cacert="/etc/etcd/ca.crt"           --endpoints="https://geliu1831master-etcd-nfs-1.int.1119-wgp.qe.rhcloud.com:2379" endpoint health
https://geliu1831master-etcd-nfs-1.int.1119-wgp.qe.rhcloud.com:2379 is healthy: successfully committed proposal: took = 986.925µs
https://geliu1831master-etcd-nfs-1:2379 is unhealthy: failed to commit proposal: context deadline exceeded
Error:  unhealthy cluster

# oc get pods  -n kube-system
NAME                                            READY     STATUS    RESTARTS   AGE
master-api-geliu1831master-etcd-nfs-1           1/1       Running   2          5h
master-controllers-geliu1831master-etcd-nfs-1   1/1       Running   2          5h
master-etcd-geliu1831master-etcd-nfs-1          1/1       Running   2          5h

# oc exec ${ETCD_POD} -c etcd -- /bin/bash -c "ETCDCTL_API=3 etcdctl \
>     --cert /etc/etcd/peer.crt \
>     --key /etc/etcd/peer.key \
>     --cacert /etc/etcd/ca.crt \
>     --endpoints $ETCD_EP \
>     snapshot save /var/lib/etcd/snapshot.db"
Error from server (BadRequest): container etcd is not valid for pod master-api-geliu1831master-etcd-nfs-1
# echo $ETCD_EP
10.0.151.209:2379
[root@geliu1831master-etcd-nfs-1 ~]# echo  ${ETCD_POD}
master-api-geliu1831master-etcd-nfs-1

I'm not sure if this is a bug. thanks

Reply:
@geliu2016 Without being able to look at the cluster, here's what I'm noticing. Your `echo ${ETCD_POD}` command returns `master-api-...`. Wouldn't it have to be `master-etcd-...` for it to be an etcd pod?

I'm not sure how `ETCD_POD` got set, but I can imagine that if it's not using an etcd pod, then any `etcdctl` commands wouldn't succeed.

Maybe it's because your cluster has "etcd" in the name? So the command to set `ETCD_POD` hit on a pod that wasn't actually an etcd pod?:

```
$ export ETCD_POD=$(oc get pods -n kube-system | grep -o -m 1 '\S*etcd\S*')
```

CC @hexfusion / @retroflexer for taking a closer look at this command

@geliu2016 in the meantime, is it possible to test on a cluster that does not have "etcd" in the name? And if the command really is the problem, we can take a look at fixing that.

Reply:
I tried it again, and the env var is not exact in some cluster, such as:

# oc get pods
NAME                                           READY     STATUS    RESTARTS   AGE
master-api-geliu311master-etcd-nfs-1           1/1       Running   2          4h
master-controllers-geliu311master-etcd-nfs-1   1/1       Running   2          4h
master-etcd-geliu311master-etcd-nfs-1          1/1       Running   2          4h

# export ETCD_POD=$(oc get pods -n kube-system | grep -o -m 1 '\S*etcd\S*')
# echo $ETCD_POD
master-api-geliu311master-etcd-nfs-1  **==> we may find the env pod is extract master-api pod,
I reset the ETCP_POD with exact one, then backup/restore both work well:**



# oc exec ${ETCD_POD} -c etcd -- /bin/bash -c "ETCDCTL_API=3 etcdctl \
>     --cert /etc/etcd/peer.crt \
>     --key /etc/etcd/peer.key \
>     --cacert /etc/etcd/ca.crt \
>     --endpoints $ETCD_EP \
>     snapshot save /var/lib/etcd/snapshot.db"
Snapshot saved at /var/lib/etcd/snapshot.db

# etcdctl snapshot restore  /var/lib/etcd.old/snapshot.db  --data-dir /var/lib/etcd/ \
>  --name geliu311master-etcd-nfs-1.int.1202-lhp.qe.rhcloud.com --initial-cluster "geliu311master-etcd-nfs-1.int.1202-lhp.qe.rhcloud.com=https://10.0.150.101:2380" --initial-cluster-token "etcd-cluster-1" --initial-advertise-peer-urls "https://10.0.150.101:2380" --skip-hash-check=true
2020-12-02 03:20:52.002931 I | mvcc: restore compact to 42280
2020-12-02 03:20:52.041995 I | etcdserver/membership: added member b3982740e78119d6 [https://10.0.150.101:2380] to cluster a47c72ed8011bc46

and after restore and sequence steps in doc, the etcd pods may start, and cluster works well.
@bergerhoffer  as memtioned above, there is a minor err in doc about env var ETCD_POD, so pls feel free to clarify it. thanks

Reply:
@geliu2016 I think the following command may be a better substitute:

export ETCD_POD=$( oc get pods -n kube-system | grep -o -m 1 '^master-etcd\S*')